File: node/test/node/stripe.test.ts
import { describe, expect, it } from 'vitest';

import { metadataForVisitor, metadataFromCurrentVisitor, mergeMetadata } from '../../src/node/stripe.js';
import { runWithVisitorId } from '../../src/node/visitor-context.js';

describe('Stripe metadata', () => {
  it('validates and trims explicit visitors', () => {
    expect(metadataForVisitor(' visitor_A-1 ')).toEqual({ cekat_visitor_id: 'visitor_A-1' });
    expect(metadataForVisitor('a')).toEqual({ cekat_visitor_id: 'a' });
    expect(metadataForVisitor('a'.repeat(128))).toEqual({ cekat_visitor_id: 'a'.repeat(128) });
    expect(metadataForVisitor('a'.repeat(129))).toEqual({});
    expect(metadataForVisitor('invalid visitor')).toEqual({});
  });

  it('reads only a valid current visitor and returns fresh metadata', () => {
    expect(runWithVisitorId(' scoped ', metadataFromCurrentVisitor)).toEqual({ cekat_visitor_id: 'scoped' });
    expect(metadataFromCurrentVisitor()).toEqual({});
    expect(metadataForVisitor('visitor')).not.toBe(metadataForVisitor('visitor'));
  });

  it('merges without mutating the merchant metadata', () => {
    const merchant = { merchant: 'keep', cekat_visitor_id: 'replace' };
    const merged = mergeMetadata(merchant, ' visitor_2 ');
    expect(merged).toEqual({ merchant: 'keep', cekat_visitor_id: 'visitor_2' });
    expect(merged).not.toBe(merchant);
    expect(merchant).toEqual({ merchant: 'keep', cekat_visitor_id: 'replace' });
    expect(mergeMetadata(merchant, 'invalid visitor')).toEqual(merchant);
    expect(mergeMetadata(merchant, 'invalid visitor')).not.toBe(merchant);
  });
});

File: go/stripe/stripe_test.go
package stripe

import (
	"context"
	"strings"
	"testing"

	cekat "golang.cekat.ai/event-sdk"
)

func TestMetadataForVisitor(t *testing.T) {
	for _, test := range []struct {
		name  string
		input string
		want  string
	}{
		{name: "trims", input: " visitor_A-1 ", want: "visitor_A-1"},
		{name: "minimum length", input: "a", want: "a"},
		{name: "maximum length", input: strings.Repeat("a", 128), want: strings.Repeat("a", 128)},
		{name: "too long", input: strings.Repeat("a", 129)},
		{name: "invalid character", input: "visitor value"},
		{name: "blank", input: " \t "},
	} {
		t.Run(test.name, func(t *testing.T) {
			metadata := MetadataForVisitor(test.input)
			if test.want == "" {
				if len(metadata) != 0 {
					t.Fatalf("MetadataForVisitor(%q) = %#v, want empty", test.input, metadata)
				}
				return
			}
			if got := metadata["cekat_visitor_id"]; got != test.want {
				t.Fatalf("MetadataForVisitor(%q) = %q, want %q", test.input, got, test.want)
			}
		})
	}
}

func TestMetadataFromContextAndMergeMetadata(t *testing.T) {
	ctx := cekat.WithVisitorID(context.Background(), " scoped_visitor ")
	fromContext := MetadataFromContext(ctx)
	if got := fromContext["cekat_visitor_id"]; got != "scoped_visitor" {
		t.Fatalf("MetadataFromContext() = %q", got)
	}
	if metadata := MetadataFromContext(context.Background()); len(metadata) != 0 {
		t.Fatalf("MetadataFromContext(empty) = %#v, want empty", metadata)
	}

	original := map[string]string{"merchant": "keep", "cekat_visitor_id": "replace"}
	merged := MergeMetadata(original, " visitor_2 ")
	if merged["merchant"] != "keep" || merged["cekat_visitor_id"] != "visitor_2" {
		t.Fatalf("MergeMetadata() = %#v", merged)
	}
	if original["cekat_visitor_id"] != "replace" {
		t.Fatalf("MergeMetadata mutated original: %#v", original)
	}
	merged["merchant"] = "changed"
	if original["merchant"] != "keep" {
		t.Fatalf("merged collection aliases original: %#v", original)
	}
	unchanged := MergeMetadata(original, "invalid value")
	if unchanged["cekat_visitor_id"] != "replace" || len(unchanged) != len(original) {
		t.Fatalf("invalid merge = %#v", unchanged)
	}
}
